### PR TITLE
Feature/rm 9504 TVPs verursachen schlechte query plans v7.0

### PR DESCRIPTION
**Pull request by martin-herwich**
Cherry pick from 8.0
https://jira.rubicon.eu/browse/RM-9504

Update: Added threshold switch between single-element parameters and table-valued parameter.

### CODE REVIEW

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs [hunk 1]
```diff
@@ -116,12 +116,12 @@ public IDbCommandBuilder CreateForSelect (
...
 
-    public IDbCommandBuilder CreateForQuery (string statement, IEnumerable<QueryParameterWithDataParameterDefinition> parametersWithType)
+    public IDbCommandBuilder CreateForQuery (string statement, IReadOnlyCollection<QueryParameterWithDataParameterDefinition> parametersWithType)
     {
       ArgumentUtility.CheckNotNull("statement", statement);
       ArgumentUtility.CheckNotNull("parametersWithType", parametersWithType);
 
-      return new QueryDbCommandBuilder(statement, parametersWithType, _sqlDialect);
+      return new SqlQueryDbCommandBuilder(statement, parametersWithType, _sqlDialect);
     }
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs]: For consistency reasons, we should stick to IEnumerable on the interface and do a ToArray when passign the parameter into SqlQueryDbCommandBuilder. I'm not opposed to one day changing the IEnumerable to IReadOnlyCollection on the factory, but we should do this wholesale instead of bit by bit.

Comment (MichaelKetting) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs]: And I just realized: I missed this on the develop-PR. We should keep the public api on the HF branch, and we should consider adjusting the API on develop.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs [hunk 2]
```diff
@@ -28,51 +31,76 @@ namespace Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders
...
   public class QueryDbCommandBuilder : DbCommandBuilder
   {
+    private static readonly ConcurrentDictionary<string, Regex> s_regexDictionary = new();
     private readonly string _statement;
-    private readonly QueryParameterWithDataParameterDefinition[] _parametersWithDefinition;
+    private readonly IReadOnlyDictionary<string, QueryParameterWithDataParameterDefinition> _parameters;
 
-    public QueryDbCommandBuilder (string statement, IEnumerable<QueryParameterWithDataParameterDefinition> parameters, ISqlDialect sqlDialect)
+
+    public QueryDbCommandBuilder (string statement, IReadOnlyCollection<QueryParameterWithDataParameterDefinition> parameters, ISqlDialect sqlDialect)
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs]: I am considering to recommend we keep this as IEnumerable, for API stability. It doesn't make a difference in the implementation and is more graceful to the callers. Should possibly also revert API change on develop.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlQueryDbCommandBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      IndexName = indexName;
+      HasIndex = hasIndex;
+    }
+
+    public string IndexName { get; }
+    public string TempTableName { get; }
+    public bool HasIndex { get; }
+  }
+
+  public SqlQueryDbCommandBuilder (string statement, IReadOnlyCollection<QueryParameterWithDataParameterDefinition> parameters, ISqlDialect sqlDialect)
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlQueryDbCommandBuilder.cs:57]: new type can and should keep IReadOnlyCollection. This prevents double calculation of the sequence.

**Review by MichaelKetting — APPROVED** [review]
I'm assuming it was a pure cherry-pick. Got some details on breaking changes in remarks we should consider but otherwise, it's done.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs (hunk 1 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs]: I would vote against reverting the API changes on develop, because this change from `IEnumerable<T>` to `IReadOnlyCollection<T>` should not break any existing caller, and even if it does, it is a trivial fix.
However, we can and probably should change all uses of `IEnumerable<T>` to `IReadOnlyCollection<T>` in `IDbCommandBuilderFactory`.

Comment (MichaelKetting) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs]: DependDB does agree about there not being any breaking usages and we only have our products that have taken a known dependency on the APIs. I'm okay with changing this towards IReadOnlyCollection in v8.0 if you check with Neon Team.

Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs]: The only call to `CreateForQuery` uses an array.
Calls to `CreateForDelete` use `ObjectIDStorageProperty.SplitValue()`, which returns an `IEnumerable<ColumnValue>` :-(
Calls to `CreateForInsert` and `CreateForUpdate` use concatenation and would have to add `.ToArray()`

So, the existing change will not break anything. But if we change the other calls, we would need to change `IRdbmsStoragePropertyDefinition` as well (for Delete) and accept trivial breaking changes in two places. The easiest course of action would be to leaven our 8.0 branch as it is, revert the API changes in 7.0, and postpone dealing with the other `IEnumerables`.

Comment (MichaelKetting) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlDbCommandBuilderFactory.cs]: As discussed, let's keep the old API also on develop until we can do a consistent change.

**Comment by martin-herwich** [comment]
Added threshold switch between single-element parameters and table-valued parameter.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Sql2016/SqlStorageObjectFactory.cs [hunk 4]
```diff
@@ -572,8 +572,14 @@ protected virtual ISqlQueryGenerator CreateSqlQueryGenerator (
...
+      var tableValuedParameterThreshold = GetTableValuedParameterThreshold();
 
-      return new SqlQueryGenerator(sqlPreparationStage, mappingResolutionStage, sqlGenerationStage);
+      return new SqlQueryGenerator(sqlPreparationStage, mappingResolutionStage, sqlGenerationStage, tableValuedParameterThreshold);
+    }
+
+    protected virtual int GetTableValuedParameterThreshold ()
+    {
+      return 1;
     }
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Sql2016/SqlStorageObjectFactory.cs]: Rick mentioned yesterday that 25 items is probably the safe break off for switching from individual values to TVP (it's not final, but very likely). Should we do this in re-motion or downstream? I'm open to fixing this in re-motion.

Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Sql2016/SqlStorageObjectFactory.cs]: Please add an xml doc here because it's rather important to explain what the value does. Including the special case if needed as well as the set default value.

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs [hunk 5]
```diff
@@ -25,11 +29,44 @@ namespace Remotion.Data.DomainObjects.Linq;
...
 {
+  private readonly int _tableValuedParameterThreshold;
+
+  public TableValuedParameterSqlCommandBuilder (int tableValuedParameterThreshold)
+  {
+    if (tableValuedParameterThreshold < 0)
+    {
+      throw new ArgumentOutOfRangeException(
+          "Threshold must be greater than 0 in order to use table-valued parameters for collections with at least that number of elements, "
+          + "or 0 to disable table-valued parameters.");
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]: please add argument name

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs [hunk 6]
```diff
@@ -25,11 +29,44 @@ namespace Remotion.Data.DomainObjects.Linq;
...
+    if (_tableValuedParameterThreshold == 0)
+    {
+      base.AppendNonEmptyCollection(collectionExpression);
+    }
+    else
+    {
+      var count = (collectionExpression.Collection as ICollection)?.Count
+          ?? (collectionExpression.Collection as ICollection<object>)?.Count
+          ?? (collectionExpression.Collection as IReadOnlyCollection<object>)?.Count
+          ?? (collectionExpression.Collection as IEnumerable)?.Cast<object>().Count();
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]: This API works with ConstantCollectionExpression.

Does it even allow IEnumerable, or only the wellknown versions of ICollection? I suggset, if it's IEnumerable, we either default to parameters or to TVP instead of iterating the IEnumerable, which is always a bad (TM) thing to do. Or we know that it's invalid and throw. This depends on the contract of ConstantCollectionExpression.

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs [hunk 7]
```diff
@@ -25,11 +29,44 @@ namespace Remotion.Data.DomainObjects.Linq;
...
   protected override void AppendNonEmptyCollection (ConstantCollectionExpression collectionExpression)
   {
     ArgumentUtility.CheckNotNull(nameof(collectionExpression), collectionExpression);
 
-    Append("SELECT [Value] FROM ");
-    AppendParameter(collectionExpression.Collection);
+    if (_tableValuedParameterThreshold == 0)
+    {
+      base.AppendNonEmptyCollection(collectionExpression);
+    }
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]: Not sure if we need the special case.
Also, the "disable" value to be 0 instead of -1 feels odd to me as far as .net convetions are concerned. If we don't need the special case, we can drop this.

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs [hunk 8]
```diff
@@ -25,11 +29,44 @@ namespace Remotion.Data.DomainObjects.Linq;
 /// </summary>
 public class TableValuedParameterSqlCommandBuilder : SqlCommandBuilder
 {
+  private readonly int _tableValuedParameterThreshold;
+
+  public TableValuedParameterSqlCommandBuilder (int tableValuedParameterThreshold)
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs:38]: please add xml doc explaining the paramter and the special cases

File: Remotion/Data/DomainObjects.UnitTests/Linq/TableValuedParameterSqlCommandBuilderTest.cs [hunk 9]
```diff
@@ -40,4 +56,63 @@ public void AppendCollection_UsesSingleParameter ()
...
+    Assert.That(parameter[0].Name, Is.EqualTo("@1"));
+    Assert.That(parameter[0].Value, Is.EqualTo(collection[0]));
+    Assert.That(parameter[1].Name, Is.EqualTo("@2"));
+    Assert.That(parameter[1].Value, Is.EqualTo(collection[1]));
+    Assert.That(parameter[2].Name, Is.EqualTo("@3"));
+    Assert.That(parameter[2].Value, Is.EqualTo(collection[2]));
+  }
+
+  [Test]
+  public void AppendCollection_ZeroThreshold_UsesMultipleParameters ()
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Linq/TableValuedParameterSqlCommandBuilderTest.cs]: please also add a test with empty collection and threshhold=0

File: Remotion/Data/DomainObjects.UnitTests/Linq/TableValuedParameterSqlCommandBuilderTest.cs [hunk 10]
```diff
@@ -25,12 +26,27 @@ namespace Remotion.Data.DomainObjects.UnitTests.Linq;
 public class TableValuedParameterSqlCommandBuilderTest
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Linq/TableValuedParameterSqlCommandBuilderTest.cs:29]: please add test for argument check in ctor

File: Remotion/Data/DomainObjects.UnitTests/Linq/SqlQueryGeneratorTest.cs [hunk 11]
```diff
@@ -50,7 +50,7 @@ public void SetUp ()
       _resolutionStageMock = new Mock<IMappingResolutionStage>(MockBehavior.Strict);
       _generationStageMock = new Mock<ISqlGenerationStage>(MockBehavior.Strict);
 
-      _sqlQueryGenerator = new SqlQueryGenerator(_preparationStageMock.Object, _resolutionStageMock.Object, _generationStageMock.Object);
+      _sqlQueryGenerator = new SqlQueryGenerator(_preparationStageMock.Object, _resolutionStageMock.Object, _generationStageMock.Object, 0);
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Linq/SqlQueryGeneratorTest.cs]: We should have a test that verifies the value is correctly injected. We can do this test later (ticket please) because it's a bit tricky to do and I don't have a good idea on how to do this quick and easy right now.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
I checked the newly changed files, assuming the rest is identical to develop

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Sql2016/SqlStorageObjectFactory.cs (hunk 4 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Sql2016/SqlStorageObjectFactory.cs]: We can use 25 here because it is a reasonable default. Downstream, we want to get the value from configuration.

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs (hunk 6 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]: `ConstantCollectionExpression` says that it represents "an ICollection, an ICollection<T> or an IReadOnlyCollection<out T>."
But the constructor parameter is an `IEnumerable`, because that is the common interface, and we need to detect whether the collection is empty.
Unfortunately, I just noticed that the soft cast to `ICollection<object>` will not work because unlike `IReadOnlyCollection<out T>`, the type parameter in `ICollection<T>` is not covariant.
So we have to fall back to `IEnumerable` anyway, unless we want to add a `Count` property to `ConstantCollectionExpression`.

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs (hunk 7 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]: I had the same thought, but on the other hand, what does a threshold of 0 mean?
We can start the allowed range at 1 instead of 0, and if someone wants to always use single-element parameters, they can use `int.MaxValue` as threshold.

File: Remotion/Data/DomainObjects.UnitTests/Linq/TableValuedParameterSqlCommandBuilderTest.cs (hunk 9 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects.UnitTests/Linq/TableValuedParameterSqlCommandBuilderTest.cs]: This is a test case in AppendCollection_Empty_UsesNoParameter.

File: Remotion/Data/DomainObjects.UnitTests/Linq/SqlQueryGeneratorTest.cs (hunk 11 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects.UnitTests/Linq/SqlQueryGeneratorTest.cs]: We could extract a method `CreateSqlCommandBuilder()` and get that via `PrivateInvoke`, right?

Comment (martin-herwich) [reply Remotion/Data/DomainObjects.UnitTests/Linq/SqlQueryGeneratorTest.cs]: Solution: we mock the `ISqlBuilderGenerationStage`, capture the `SqlCommandBuilder` and assert the new `TableValuedParameterThreshold` property.

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs (hunk 6 above)
Comment (martin-herwich) [reply Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]: TODO: UnitTests for the different branches.

File: Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Sql2016/SqlStorageObjectFactory.cs [hunk 12]
```diff
@@ -572,8 +572,18 @@ protected virtual ISqlQueryGenerator CreateSqlQueryGenerator (
...
       var mappingResolutionStage = ObjectFactory.Create<DefaultMappingResolutionStage>(ParamList.Create(resolver, generator));
       var sqlGenerationStage = ObjectFactory.Create<ExtendedSqlGenerationStage>(ParamList.Empty);
+      var tableValuedParameterThreshold = GetTableValuedParameterThreshold();
 
-      return new SqlQueryGenerator(sqlPreparationStage, mappingResolutionStage, sqlGenerationStage);
+      return new SqlQueryGenerator(sqlPreparationStage, mappingResolutionStage, sqlGenerationStage, tableValuedParameterThreshold);
+    }
+
+    /// <summary>
+    /// Gets the number of elements in a collection that triggers the use of a table-valued parameter instead of single-element parameters in
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Sql2016/SqlStorageObjectFactory.cs:581]: Please also add a note to the jira ticket for the TVP feature (even though the feature is already released) so there is a hint about this.

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs [hunk 13]
```diff
@@ -25,11 +30,49 @@ namespace Remotion.Data.DomainObjects.Linq;
...
+  /// <summary>
+  /// Creates a new <see cref="SqlCommandBuilder"/> instance that uses a table-valued parameter (instead of single-element parameters) for a collection that has at least
+  /// <paramref name="tableValuedParameterThreshold"/> elements.
+  /// </summary>
+  /// <exception cref="ArgumentOutOfRangeException">If <paramref name="tableValuedParameterThreshold"/> is 0 or negative.</exception>
+  public TableValuedParameterSqlCommandBuilder (int tableValuedParameterThreshold)
+  {
+    if (tableValuedParameterThreshold <= 0)
+    {
+      throw new ArgumentOutOfRangeException(nameof(tableValuedParameterThreshold), "Threshold must be greater than 0");
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]:
```suggestion
      throw new ArgumentOutOfRangeException(nameof(tableValuedParameterThreshold), "Threshold must be greater than 0.");
```

File: Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs [hunk 14]
```diff
@@ -25,11 +30,49 @@ namespace Remotion.Data.DomainObjects.Linq;
...
+
+  public int TableValuedParameterThreshold { get; }
+
   protected override void AppendNonEmptyCollection (ConstantCollectionExpression collectionExpression)
   {
     ArgumentUtility.CheckNotNull(nameof(collectionExpression), collectionExpression);
 
-    Append("SELECT [Value] FROM ");
-    AppendParameter(collectionExpression.Collection);
+    int GetCount (object enumerable)
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects/Linq/TableValuedParameterSqlCommandBuilder.cs]:
```suggestion
    static int GetCount (object enumerable)
```

File: Remotion/Data/DomainObjects.UnitTests/Factories/BaseConfiguration.cs [hunk 15]
```diff
@@ -122,7 +123,7 @@ private FakeStorageSettingsFactory CreateStorageSettings ()
 
       var deferredStorageSettings = new DeferredStorageSettings(fakeStorageObjectFactoryFactory, fakeStorageSettingsFactoryResolver);
 
-      var sqlStorageObjectFactory = new SqlStorageObjectFactory(
+      var sqlStorageObjectFactory = new TestSqlStorageObjectFactory(
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Factories/BaseConfiguration.cs:126]: Yes!

**Review by MichaelKetting — APPROVED** [review]
Assuming this is only the new test SQL storageobjfactory four or so new changed files?  Mobil GitHub app isn't so great for review
